Update joint_trajectory_controller_impl.h for issue #48

onionsflying opened the pull request:
fix the issue #48

onionsflying committed:
Update joint_trajectory_controller_impl.h

fix the issue #48

mathias-luedtke commented:
This is basically the same patch @ipa-fxm  did in https://github.com/ipa-fxm/ros_controllers/commit/7b5a88a89010fcbda1cc9819871d8cb3332b61e9, but it [does not work properly](https://github.com/ros-controls/ros_controllers/issues/48#issuecomment-81122306).

onionsflying committed:
fix a bug about the conflict between action aborting and setHoldPosition method

mathias-luedtke commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -193,11 +193,16 @@ checkPathTolerances(const typename Segment::State& state_error,
     
+    // Sleep until the goal is aborted
+    ros::Duration action_monitor_period_2x=action_monitor_period_.operator *(2);
+    action_monitor_period_2x.sleep();
```
sleeping in the RT update cycle is a no-go.

onionsflying replied:
ok, understand. It turns out that the [update](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L374) loop after calling [setHoldPosition](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L671) can destroy the [goal_handle_timer_](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L567). So I make update-thread sleep for awhile until the timer run at least one more time for aborting the action. There should be many solutions without sleeping.

onionsflying committed:
call setHoldPosition(time) outside the RT update

onionsflying commented:
I think [setHoldPosition](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L671) should be called outside the RT update. It consumes about 0.3 ms, which is a little long for RT.

onionsflying commented:
Hi, @ipa-mdl ! Would you like to review the new commit? If you have any suggestion, please tell me.

mathias-luedtke commented:
I am not convinced that this is a proper fix, it might even introduce a race condition with multi-threaded spinners.

> I think setHoldPosition should be called outside the RT update. It consumes about 0.3 ms, which is a little long for RT.

How did you measure this?
And why are 0.3ms too long for RT?

> If you have any suggestion, please tell me.

I think it would be great to have a unit test for 0d3de78ef8b19b54beb4a6b007c7e179d57aa5e0.
Their might be a bug somewehe underneath that leads to problem with lost goals.

onionsflying commented:
I just added some codes e.g. ROS_DEBUG_NAMED(name, "msg, %f", ros::Time::now().toSec()) into [joint_trajectory_controller_imp.h](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h) and [realtime_server_goal_handle.h](https://github.com/ros-controls/realtime_tools/blob/indigo-devel/include/realtime_tools/realtime_server_goal_handle.h) . Then I could get the informations that I mentioned. It's obviously not a strict test. I thought a RT cycle should take less than 1ms. And 0.3ms could be a big part, that might make the whole cycle take more than 1ms. So I wanna insist on my thoughts, since I couldn't imagine the race condition. [cancelCB(GoalHandle gh)](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L583) should be a method without race condition. Have you seen some differences between actionResultCB in 0a7f330 and [cancelCB](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L583) from the perspective of race condition?

I agree that there might be a bug somewhere. Thus I'm not able to explain why the timer [goal_handle_timer_](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L567) stops running abnormally when we patched the code in a way like 0d3de78 . (I know when the timer stops, because I added a line _ROS_DEBUG_NAMED("realtime_tools", "update runNonRealtime");_ in [runNonRealtime(const ros::TimerEvent &te)](https://github.com/ros-controls/realtime_tools/blob/indigo-devel/include/realtime_tools/realtime_server_goal_handle.h#L97) .)

mathias-luedtke commented:
> I couldn't imagine the race condition

Your code stops the motions without checking the goal matching the aborted one.
However, I don't think that adding a subscriber is a proper fix.

> I just added some codes e.g. ROS_DEBUG_NAMED(name, "msg, %f", ros::Time::now().toSec()) into 

This not a very precise measurement..

>  I thought a RT cycle should take less than 1ms. And 0.3ms could be a big part, that might make the whole cycle take more than 1ms.

In our context RT is about determinism and non-blocking behavior.
The actual cycle time is bound to the application.

> I agree that there might be a bug somewhere. Thus I'm not able to explain why the timer goal_handle_timer_ stops running abnormally when we patched the code in a way like 0d3de78

That's why it would be great to have a unit test for this case.
It might even help to do proper time measurements.

bmagyar commented:
Does this PR still hold relevance? Can I close it? Should we retarget/rework for melodic?
